File: src/bf16/bf16_math_one_arg.c
// SPDX-License-Identifier LGPL-3.0-or-later
/*
  This file is part of LPF, a Low Precision helper for Fortran
  Copyright (C) 2025 Martin Koehler

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 3 of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with this program; if not, write to the Free Software Foundation,
  Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <math.h>

#include "bf16_helper.h"

/*
 * Function: acos ( C: acosf )
 */
HIDDEN void __bf16_helper_acos(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) acosf ((float) _a.bf16);
}

/*
 * Function: acosh ( C: acoshf )
 */
HIDDEN void __bf16_helper_acosh(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) acoshf ((float) _a.bf16);
}

/*
 * Function: asin ( C: asinf )
 */
HIDDEN void __bf16_helper_asin(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) asinf ((float) _a.bf16);
}

/*
 * Function: asinh ( C: asinhf )
 */
HIDDEN void __bf16_helper_asinh(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) asinhf ((float) _a.bf16);
}

/*
 * Function: atan ( C: atanf )
 */
HIDDEN void __bf16_helper_atan(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) atanf ((float) _a.bf16);
}

/*
 * Function: atanh ( C: atanhf )
 */
HIDDEN void __bf16_helper_atanh(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) atanhf ((float) _a.bf16);
}

/*
 * Function: bessel_j0 ( C: j0f )
 */
HIDDEN void __bf16_helper_bessel_j0(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) j0f ((float) _a.bf16);
}

/*
 * Function: bessel_j1 ( C: j1f )
 */
HIDDEN void __bf16_helper_bessel_j1(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) j1f ((float) _a.bf16);
}

/*
 * Function: bessel_y0 ( C: y0f )
 */
HIDDEN void __bf16_helper_bessel_y0(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) y0f ((float) _a.bf16);
}

/*
 * Function: bessel_y1 ( C: y1f )
 */
HIDDEN void __bf16_helper_bessel_y1(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) y1f ((float) _a.bf16);
}

/*
 * Function: ceiling ( C: ceilf )
 */
HIDDEN void __bf16_helper_ceiling(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) ceilf ((float) _a.bf16);
}

/*
 * Function: cos ( C: cosf )
 */
HIDDEN void __bf16_helper_cos(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) cosf ((float) _a.bf16);
}

/*
 * Function: cosh ( C: coshf )
 */
HIDDEN void __bf16_helper_cosh(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) coshf ((float) _a.bf16);
}

/*
 * Function: erf ( C: erff )
 */
HIDDEN void __bf16_helper_erf(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) erff ((float) _a.bf16);
}

/*
 * Function: erfc ( C: erfcf )
 */
HIDDEN void __bf16_helper_erfc(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) erfcf ((float) _a.bf16);
}

/*
 * Function: exp ( C: expf )
 */
HIDDEN void __bf16_helper_exp(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) expf ((float) _a.bf16);
}

/*
 * Function: floor ( C: floorf )
 */
HIDDEN void __bf16_helper_floor(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) floorf ((float) _a.bf16);
}

/*
 * Function: gamma ( C: tgammaf )
 */
HIDDEN void __bf16_helper_gamma(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) tgammaf ((float) _a.bf16);
}

/*
 * Function: log ( C: logf )
 */
HIDDEN void __bf16_helper_log(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) logf ((float) _a.bf16);
}

/*
 * Function: log10 ( C: log10f )
 */
HIDDEN void __bf16_helper_log10(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) log10f ((float) _a.bf16);
}

/*
 * Function: log_gamma ( C: lgammaf )
 */
HIDDEN void __bf16_helper_log_gamma(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) lgammaf ((float) _a.bf16);
}

/*
 * Function: sin ( C: sinf )
 */
HIDDEN void __bf16_helper_sin(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) sinf ((float) _a.bf16);
}

/*
 * Function: sinh ( C: sinhf )
 */
HIDDEN void __bf16_helper_sinh(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) sinhf ((float) _a.bf16);
}

/*
 * Function: sqrt ( C: sqrtf )
 */
HIDDEN void __bf16_helper_sqrt(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) sqrtf ((float) _a.bf16);
}

/*
 * Function: tan ( C: tanf )
 */
HIDDEN void __bf16_helper_tan(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) tanf ((float) _a.bf16);
}

/*
 * Function: tanh ( C: tanhf )
 */
HIDDEN void __bf16_helper_tanh(int16_t *out, int16_t in)
{
    bf16_handler_t *r = (bf16_handler_t * ) out;
    bf16_handler_t _a = { .i16 = in};
    r->bf16 = (__bf16) tanhf ((float) _a.bf16);
}

File: src/bf16/bf16_math_one_arg.f90
!  SPDX-License-Identifier: LGPL-3.0-or-later
!
!  This file is part of LPF, a Low Precision helper for Fortran
!  Copyright (C) 2025 Martin Koehler
!
!  This program is free software; you can redistribute it and/or
!  modify it under the terms of the GNU Lesser General Public
!  License as published by the Free Software Foundation; either
!  version 3 of the License, or (at your option) any later version.
!
!  This program is distributed in the hope that it will be useful,
!  but WITHOUT ANY WARRANTY; without even the implied warranty of
!  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
!  Lesser General Public License for more details.
!
!  You should have received a copy of the GNU Lesser General Public License
!  along with this program; if not, write to the Free Software Foundation,
!  Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
!

submodule (lpf_bf16) lpf_bf16_math_one_arg
    use iso_c_binding
    use iso_fortran_env, only: real32, real64
    implicit none

    interface
        pure subroutine helper_bf16_acos(out, in) bind(c, name="__bf16_helper_acos")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_acosh(out, in) bind(c, name="__bf16_helper_acosh")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_asin(out, in) bind(c, name="__bf16_helper_asin")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_asinh(out, in) bind(c, name="__bf16_helper_asinh")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_atan(out, in) bind(c, name="__bf16_helper_atan")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_atanh(out, in) bind(c, name="__bf16_helper_atanh")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_bessel_j0(out, in) bind(c, name="__bf16_helper_bessel_j0")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_bessel_j1(out, in) bind(c, name="__bf16_helper_bessel_j1")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_bessel_y0(out, in) bind(c, name="__bf16_helper_bessel_y0")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_bessel_y1(out, in) bind(c, name="__bf16_helper_bessel_y1")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_ceiling(out, in) bind(c, name="__bf16_helper_ceiling")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_cos(out, in) bind(c, name="__bf16_helper_cos")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_cosh(out, in) bind(c, name="__bf16_helper_cosh")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_erf(out, in) bind(c, name="__bf16_helper_erf")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_erfc(out, in) bind(c, name="__bf16_helper_erfc")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_exp(out, in) bind(c, name="__bf16_helper_exp")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_floor(out, in) bind(c, name="__bf16_helper_floor")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_gamma(out, in) bind(c, name="__bf16_helper_gamma")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_log(out, in) bind(c, name="__bf16_helper_log")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_log10(out, in) bind(c, name="__bf16_helper_log10")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_log_gamma(out, in) bind(c, name="__bf16_helper_log_gamma")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_sin(out, in) bind(c, name="__bf16_helper_sin")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_sinh(out, in) bind(c, name="__bf16_helper_sinh")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_sqrt(out, in) bind(c, name="__bf16_helper_sqrt")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_tan(out, in) bind(c, name="__bf16_helper_tan")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
        pure subroutine helper_bf16_tanh(out, in) bind(c, name="__bf16_helper_tanh")
            use, intrinsic :: iso_c_binding
            integer(c_int16_t), intent(out) :: out
            integer(c_int16_t), intent(in), value :: in
        end subroutine
    end interface
contains
    module elemental function acos_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_acos(out%value, in%value)
    end function
    module elemental function acosh_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_acosh(out%value, in%value)
    end function
    module elemental function asin_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_asin(out%value, in%value)
    end function
    module elemental function asinh_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_asinh(out%value, in%value)
    end function
    module elemental function atan_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_atan(out%value, in%value)
    end function
    module elemental function atanh_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_atanh(out%value, in%value)
    end function
    module elemental function bessel_j0_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_bessel_j0(out%value, in%value)
    end function
    module elemental function bessel_j1_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_bessel_j1(out%value, in%value)
    end function
    module elemental function bessel_y0_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_bessel_y0(out%value, in%value)
    end function
    module elemental function bessel_y1_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_bessel_y1(out%value, in%value)
    end function
    module elemental function ceiling_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_ceiling(out%value, in%value)
    end function
    module elemental function cos_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_cos(out%value, in%value)
    end function
    module elemental function cosh_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_cosh(out%value, in%value)
    end function
    module elemental function erf_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_erf(out%value, in%value)
    end function
    module elemental function erfc_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_erfc(out%value, in%value)
    end function
    module elemental function exp_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_exp(out%value, in%value)
    end function
    module elemental function floor_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_floor(out%value, in%value)
    end function
    module elemental function gamma_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_gamma(out%value, in%value)
    end function
    module elemental function log_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_log(out%value, in%value)
    end function
    module elemental function log10_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_log10(out%value, in%value)
    end function
    module elemental function log_gamma_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_log_gamma(out%value, in%value)
    end function
    module elemental function sin_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_sin(out%value, in%value)
    end function
    module elemental function sinh_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_sinh(out%value, in%value)
    end function
    module elemental function sqrt_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_sqrt(out%value, in%value)
    end function
    module elemental function tan_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_tan(out%value, in%value)
    end function
    module elemental function tanh_bf16(in) result(out)
            type(BF16) :: out
            type(BF16), intent(in) :: in

            call helper_bf16_tanh(out%value, in%value)
    end function
end submodule
